File: benchmark/live/budget.py
"""benchmark/live/budget.py — budget enforcement for live dispatch (DV0e).

Two gates (mandate / AC-8):

  (a) PRE-FLIGHT estimate, before stage 1: project the whole-run cost via the REAL
      plugin script `skills/estimation-methodology/scripts/estimate-calc.py`. If the
      projection exceeds --budget, decline BEFORE any dispatch (dispatch nothing).

  (b) RUNNING-TALLY abort, before each stage: maintain spent_usd; before dispatching
      stage k, if spent_usd + next_stage_estimate > budget, abort (dispatch nothing
      further) and let the caller write a partial record.

Neither gate ever fabricates spend — `spent_usd` only accumulates REAL per-stage
costs that the capture layers actually returned (a stage whose cost came back None
contributes 0 to the running tally but flags the record live_partial).
"""

from __future__ import annotations

import json
import subprocess
import sys

# Per-stage expected token budget used ONLY for cost ESTIMATION (not real usage).
# Conservative round numbers; the real spend is measured per stage by dispatch.py.
PER_STAGE_EXPECTED_TOKENS = 40_000

# The 10-stage full pipeline (PL -> ... -> ST) the live WITH-path runs.
PIPELINE_STAGES = ("PL", "AR", "TL", "DV", "DR", "SR", "QA", "DC", "FN", "ST")


class BudgetExceeded(RuntimeError):
    """Raised when the pre-flight projection alone exceeds the budget (no dispatch)."""


def estimate_stage_cost(
    *,
    tokens: int = PER_STAGE_EXPECTED_TOKENS,
    model: str = "opus",
    retry_complexity: str = "high",
    estimate_calc_path: str,
    runner=None,
) -> float:
    """Project one stage's USD cost via the REAL estimate-calc.py `ai_cost` chain.

    `runner` is an injectable callable(argv: list[str]) -> str returning stdout JSON;
    defaults to a subprocess call to the real script. Tests inject a fake to avoid a
    subprocess (and to exercise specific cost sequences). Returns the script's
    `ai_cost.usd` float. On any parse/exec failure, returns 0.0 defensively — a
    failed ESTIMATE must never block a run on its own (the running-tally gate uses
    REAL spend; the pre-flight gate treats an unparseable estimate as "unknown/0").
    """
    argv = [
        sys.executable,
        estimate_calc_path,
        "--tokens",
        str(tokens),
        "--model",
        model,
        "--retry-complexity",
        retry_complexity,
    ]
    try:
        if runner is not None:
            stdout = runner(argv)
        else:
            stdout = subprocess.run(
                argv, capture_output=True, text=True, check=True
            ).stdout
        data = json.loads(stdout)
        usd = data.get("ai_cost", {}).get("usd")
        return float(usd) if usd is not None else 0.0
    except (subprocess.SubprocessError, json.JSONDecodeError, ValueError, KeyError):
        return 0.0


def preflight_projection(
    *,
    stage_count: int,
    estimate_calc_path: str,
    runner=None,
) -> float:
    """Project the whole run cost = per-stage estimate * number of stages."""
    per_stage = estimate_stage_cost(
        estimate_calc_path=estimate_calc_path, runner=runner
    )
    return per_stage * stage_count


def assert_preflight_within_budget(
    *,
    budget: float,
    stage_count: int,
    estimate_calc_path: str,
    runner=None,
) -> float:
    """Gate (a). Raise BudgetExceeded if the projection exceeds budget; else return it.

    Dispatch NOTHING on failure — the caller catches and exits non-zero before any
    `claude agents run`.
    """
    projection = preflight_projection(
        stage_count=stage_count,
        estimate_calc_path=estimate_calc_path,
        runner=runner,
    )
    if projection > budget:
        raise BudgetExceeded(
            f"pre-flight projection ${projection:.4f} exceeds budget "
            f"${budget:.4f} for {stage_count} stages; dispatching nothing"
        )
    return projection


class RunningTally:
    """Gate (b). Accumulates REAL per-stage spend and decides if the next stage fits.

    Usage per stage:
        if not tally.can_afford(next_stage_estimate):
            abort -> write partial record
        ... dispatch stage, get real cost ...
        tally.add(real_cost_or_None)
    """

    def __init__(self, budget: float) -> None:
        self.budget = budget
        self.spent_usd = 0.0

    def can_afford(self, next_stage_estimate: float) -> bool:
        """True iff dispatching the next stage would NOT breach the cap."""
        return (self.spent_usd + next_stage_estimate) <= self.budget

    def add(self, real_cost: float | None) -> None:
        """Add a stage's REAL measured cost. None (unmeasured) adds nothing.

        Never fabricates: an unmeasured stage contributes 0 to the tally; the record
        is independently flagged live_partial by the dispatcher.
        """
        if real_cost is not None:
            self.spent_usd += float(real_cost)


__all__ = [
    "PER_STAGE_EXPECTED_TOKENS",
    "PIPELINE_STAGES",
    "BudgetExceeded",
    "estimate_stage_cost",
    "preflight_projection",
    "assert_preflight_within_budget",
    "RunningTally",
]

File: benchmark/harness/Sources/BenchmarkLive/Budget.swift
// Budget.swift — budget enforcement for live dispatch, port of
// benchmark/live/budget.py.
//
// Two gates:
//   (a) PRE-FLIGHT estimate before stage 1: projection > budget -> decline
//       BEFORE any dispatch (rc=2 at the Dispatch layer).
//   (b) RUNNING-TALLY abort before EACH stage: spent + next_estimate > budget ->
//       abort, partial record with live_partial=true written BEFORE rc=4 (D6).
//
// Neither gate fabricates spend — the tally only accumulates REAL per-stage
// costs the capture layers actually returned (None adds nothing).

import BenchmarkKit
import Foundation

public enum Budget {
    /// Per-stage expected token budget used ONLY for cost ESTIMATION.
    public static let perStageExpectedTokens = 40_000

    /// The 10-stage full pipeline (PL -> ... -> ST) the live WITH-path runs.
    /// SR is IN the live pipeline (AR delta D3) even when a given worktask
    /// drops it as a stage.
    ///
    /// ADR (OI-3, branch a — AR delta D3): SR is a GENUINE, intended stage of
    /// the live benchmark pipeline, distinct from DR. The SR prompt does
    /// security-checklist work (sandbox-path/secrets/input-validation/network/
    /// shell review); the DR prompt does correctness + SwiftUI-quality review —
    /// verified non-overlapping (benchmark/live/prompts/{sr,dr}.txt), so
    /// measuring SR as its own ~23% cost stage is correct, NOT double-work.
    /// Any prior text calling SR "folded into DR/DV review" for the LIVE
    /// pipeline is a stale mis-generalization of a WORKTASK-level stage-set
    /// choice (a run may drop a standalone SR *review task*); it does NOT mean
    /// the live pipeline folds SR. Do not remove SR here without re-opening that
    /// decision — it would shift measured per-run cost baselines.
    public static let pipelineStages = ["PL", "AR", "TL", "DV", "DR", "SR",
                                        "QA", "DC", "FN", "ST"]

    public struct BudgetExceeded: Error, CustomStringConvertible {
        public let description: String
    }

    /// Project one stage's USD cost via the REAL estimate-calc.py `ai_cost`
    /// chain. `runner` is injectable ((argv) -> String stdout); defaults to a
    /// subprocess call. On ANY parse/exec failure returns 0.0 defensively — a
    /// failed ESTIMATE must never block a run on its own.
    public static func estimateStageCost(
        tokens: Int = perStageExpectedTokens,
        model: String = "opus",
        retryComplexity: String = "high",
        estimateCalcPath: String,
        runner: (([String]) throws -> String)? = nil
    ) -> Double {
        let argv = ["python3", estimateCalcPath,
                    "--tokens", String(tokens),
                    "--model", model,
                    "--retry-complexity", retryComplexity]
        let stdout: String
        if let runner {
            guard let out = try? runner(argv) else { return 0.0 }
            stdout = out
        } else {
            let r = Subprocess.run(argv)
            guard r.exitCode == 0 else { return 0.0 }
            stdout = r.stdout
        }
        guard let parsed = try? JSONParser.parse(stdout),
              let usd = parsed["ai_cost"]?["usd"]?.doubleValue else { return 0.0 }
        return usd
    }

    /// Whole-run projection = per-stage estimate * stage count.
    public static func preflightProjection(
        stageCount: Int, estimateCalcPath: String,
        runner: (([String]) throws -> String)? = nil
    ) -> Double {
        estimateStageCost(estimateCalcPath: estimateCalcPath, runner: runner)
            * Double(stageCount)
    }

    /// Gate (a). Throws BudgetExceeded if the projection exceeds budget; else
    /// returns the projection. Dispatch NOTHING on failure.
    @discardableResult
    public static func assertPreflightWithinBudget(
        budget: Double, stageCount: Int, estimateCalcPath: String,
        runner: (([String]) throws -> String)? = nil
    ) throws -> Double {
        let projection = preflightProjection(
            stageCount: stageCount, estimateCalcPath: estimateCalcPath, runner: runner)
        if projection > budget {
            throw BudgetExceeded(description: String(
                format: "pre-flight projection $%.4f exceeds budget $%.4f for %d stages; dispatching nothing",
                projection, budget, stageCount))
        }
        return projection
    }

    /// Gate (b). Accumulates REAL per-stage spend; decides if the next stage fits.
    public final class RunningTally {
        public let budget: Double
        public private(set) var spentUSD: Double = 0.0

        public init(budget: Double) { self.budget = budget }

        /// True iff dispatching the next stage would NOT breach the cap.
        public func canAfford(_ nextStageEstimate: Double) -> Bool {
            spentUSD + nextStageEstimate <= budget
        }

        /// Add a stage's REAL measured cost. nil (unmeasured) adds nothing —
        /// never fabricates; the record is independently flagged live_partial.
        public func add(_ realCost: Double?) {
            if let realCost { spentUSD += realCost }
        }
    }
}
